### PR TITLE
Refactor cross platform Interop types into System.Runtime.InteropServices.PInvoke.dll

### PR DESCRIPTION
**Pull request by botaberg**
Only a subset of the functionality in System.Runtime.InteropServices.dll can be supported cross-platform, so it is necessary to split its contract and partial facade assemblies.

System.Runtime.InteropServices.PInvoke.dll will be the new cross-platform contract.

System.Runtime.InteropServices.dll will be the "legacy" contract, available on Windows. Referencing it will allow an app to use the same set of APIs as before, including all of the COM and Windows-specific APIs. It will contain a reference to System.Runtime.InteropServices.PInvoke.dll, and type forward the cross-platform types there.

System.Runtime.InteropServices.Marshal is a special case. Some of its methods will be supported cross-platform, and moved to a new type - System.Runtime.InteropServices.PInvokeMarshal. Some of the methods' names have been changed, such as AllocCoTaskMem -> AllocateMemory. The forwarding of PInvokeMarshal methods is done using calls in the "implementation" part of the partial facade - see PInvokeMarshal.cs.

GCHandle and GCHandleType have been moved to System.Runtime.dll as part of this general refactoring - they don't have much to do with Interop.

Concerns:
-I am not sure when to update the dependencies in the project.json files. Currently the NuGet packages for the updated assemblies are not available to be referenced. In the future we may need to replace the hard project references with NuGet dependencies.

### CODE REVIEW

**Comment by botaberg** [comment]
@yizhang82 @weshaggard @ericstj @mellinoe Please take a look

**Comment by mikedn** [comment]
> GCHandle and GCHandleType have been moved to System.Runtime.dll as part of this general refactoring - they don't have much to do with Interop.

But are they the only types/members that aren't about interop? What's "interop" about `AllocateMemory` for example? Managed code can very well allocate native memory and use it for non-interop purposes.

**Comment by yizhang82** [comment]
@mikedn Typically, managed applications does not need to allocate unmanaged/native memory - the need usually comes from interop requirements. Another thing about AllocateMemory is that it is part of the contract between .NET/CLR marshaller and native code - they need to use the exact same allocate/deallocate memory helper as a contract.

File: src/System.Runtime.InteropServices.PInvoke/src/System.Runtime.InteropServices.PInvoke.builds [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <Project Include="System.Runtime.InteropServices.PInvoke.csproj">
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices.PInvoke/src/System.Runtime.InteropServices.PInvoke.builds]: This is just a facade correct? If so you shouldn't need all the different OS configurations.

File: src/System.Runtime.InteropServices.PInvoke/src/project.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "frameworks": {
+    "dnxcore50": {
+        "dependencies": {
+            "Microsoft.TargetingPack.Private.CoreCLR": "1.0.0-rc2-23520"
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices.PInvoke/src/project.json]: For consistency use "rc2-23607" here as well.

File: src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj [hunk 3]
```diff
@@ -6,17 +6,25 @@
          to its own obsolete API. -->
     <NoWarn>618</NoWarn>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <AssemblyVersion>4.0.21.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj]: Due to the number of internal TFS build breaks any version bumps like this need to be made on the TFS side. So please continue the review here but apply them your changes to TFS and check-in the changes from there.

File: src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj [hunk 4]
```diff
@@ -6,17 +6,25 @@
...
-    <Compile Include="System.Runtime.InteropServices.Manual.cs" />
   </ItemGroup>
   <ItemGroup>
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="../../System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.csproj">
+      <Private>false</Private>
+    </ProjectReference>
+    <ProjectReference Include="../../System.Runtime/ref/System.Runtime.csproj">
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj]: Why does this need to be a project reference to System.Runtime? It should just be a package reference in project.json.

File: src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj [hunk 5]
```diff
@@ -6,17 +6,25 @@
          to its own obsolete API. -->
     <NoWarn>618</NoWarn>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <AssemblyVersion>4.0.21.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
     <OutputType>Library</OutputType>
     <PackageTargetFramework>dotnet5.4</PackageTargetFramework>
```
Comment (weshaggard) [inline src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.csproj]: @ericstj I assume this needs to be bumped up to dotnet5.6, correct?

File: src/System.Runtime/ref/System.Runtime.csproj [hunk 6]
```diff
@@ -3,7 +3,7 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <AssemblyVersion>4.0.21.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
```
Comment (weshaggard) [inline src/System.Runtime/ref/System.Runtime.csproj]: Why did you change the version of System.Runtime?

File: src/System.Runtime.InteropServices.PInvoke/src/System.Runtime.InteropServices.PInvoke.builds (hunk 1 above)
Comment (botaberg) [inline src/System.Runtime.InteropServices.PInvoke/src/System.Runtime.InteropServices.PInvoke.builds]: @weshaggard This is the same as it is in System.Runtime.InteropServices.builds. It is a full facade on net46 with the exception of the redirection calls for PInvokeMarshal, and a partial facade on netcore50aot and the empty target group.

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    {
+        public DllNotFoundException() { }
+        public DllNotFoundException(string message) { }
+        public DllNotFoundException(string message, System.Exception inner) { }
+    }
+}
+namespace System.Runtime.InteropServices
+{
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public partial struct ArrayWithOffset
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: Why do we need this type here?

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+    public sealed partial class InAttribute : System.Attribute
+    {
+        public InAttribute() { }
+    }
+    public static partial class PInvokeMarshal
+    {
+        public static readonly int SystemDefaultCharSize;
+        public static readonly int SystemMaxDBCSCharSize;
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr AllocateMemory(int cb) { return default(System.IntPtr); }
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: Since it's a new type, can we fix some of the naming guideline issues? e.g. rename cb to sizeInBytes (or something like that).

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        public static void DestroyStructure(System.IntPtr ptr, System.Type structuretype) { }
+        [System.Security.SecurityCriticalAttribute]
+        public static void DestroyStructure<T>(System.IntPtr ptr) { }
+        [System.Security.SecurityCriticalAttribute]
+        public static void FreeMemory(System.IntPtr ptr) { }
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        [System.ObsoleteAttribute("GetDelegateForFunctionPointer(IntPtr, Type) may be unavailable in future releases. Instead, use GetDelegateForFunctionPointer<T>(IntPtr). For more info, go to http://go.microsoft.com/fwlink/?LinkID=296521")]
+        [System.Security.SecurityCriticalAttribute]
+        public static System.Delegate GetDelegateForFunctionPointer(System.IntPtr ptr, System.Type t) { return default(System.Delegate); }
+        [System.Security.SecurityCriticalAttribute]
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: Can we rename t to "delegateType"?

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        public static void PtrToStructure(System.IntPtr ptr, object structure) { }
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        [System.ObsoleteAttribute("PtrToStructure(IntPtr, Type) may be unavailable in future releases. Instead, use PtrToStructure<T>(IntPtr). For more info, go to http://go.microsoft.com/fwlink/?LinkID=296513")]
+        [System.Security.SecurityCriticalAttribute]
+        public static object PtrToStructure(System.IntPtr ptr, System.Type structureType) { return default(object); }
+        [System.Security.SecurityCriticalAttribute]
+        public static T PtrToStructure<T>(System.IntPtr ptr) { return default(T); }
+        [System.Security.SecurityCriticalAttribute]
+        public static void PtrToStructure<T>(System.IntPtr ptr, T structure) { }
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: We should pick one name for these parameters "structure" or "structureType". You are currently using both of these and also "t"

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+        [System.Security.SecurityCriticalAttribute]
+        public static byte ReadByte(System.IntPtr ptr) { return default(byte); }
+        [System.Security.SecurityCriticalAttribute]
+        public static byte ReadByte(System.IntPtr ptr, int ofs) { return default(byte); }
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: "ofs" should be "offset"?

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        [System.ObsoleteAttribute("UnsafeAddrOfPinnedArrayElement(Array, Int32) may be unavailable in future releases. Instead, use UnsafeAddrOfPinnedArrayElement<T>(T[], Int32). For more info, go to http://go.microsoft.com/fwlink/?LinkID=296517")]
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr UnsafeAddrOfPinnedArrayElement(System.Array arr, int index) { return default(System.IntPtr); }
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr UnsafeAddrOfPinnedArrayElement<T>(T[] arr, int index) { return default(System.IntPtr); }
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        [System.ObsoleteAttribute("WriteInt32(Object, Int32, Int32) may be unavailable in future releases.")]
+        [System.Security.SecurityCriticalAttribute]
+        public static void WriteInt32(object ptr, int ofs, int val) { ptr = default(object); }
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: value?

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr ReadIntPtr(System.IntPtr ptr) { return default(System.IntPtr); }
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr ReadIntPtr(System.IntPtr ptr, int ofs) { return default(System.IntPtr); }
+        [System.ComponentModel.EditorBrowsableAttribute((System.ComponentModel.EditorBrowsableState)(1))]
+        [System.ObsoleteAttribute("ReadIntPtr(Object, Int32) may be unavailable in future releases.")]
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr ReadIntPtr(object ptr, int ofs) { return default(System.IntPtr); }
+        [System.Security.SecurityCriticalAttribute]
+        public static System.IntPtr ReallocateMemory(System.IntPtr pv, int cb) { return default(System.IntPtr); }
```
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: pv, cb

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs (hunk 7 above)
Comment (botaberg) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: ArrayWithOffset is only used in an Interop/PInvoke context.

File: src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs (hunk 10 above)
Comment (botaberg) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: The naming makes sense to me. We use `<structureType>` when the structure parameter is of type System.Type, and `<structure>` when it is of type System.Object or generic <T>.

As for `<t>`, I can rename it to `<type>` perhaps.

Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/ref/System.Runtime.InteropServices.PInvoke.cs]: ok, sounds good.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static System.IntPtr GetFunctionPointerForDelegate(System.Delegate d)
+        {
+            return Marshal.GetFunctionPointerForDelegate(d);
+        }
+        public static System.IntPtr GetFunctionPointerForDelegate<TDelegate>(TDelegate d)
+        {
+            return Marshal.GetFunctionPointerForDelegate<TDelegate>(d);
+        }
+        public static int GetLastWin32Error()
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: I think we should name it GetLastError instead of GetLastWin32Error

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static T PtrToStructure<T>(System.IntPtr ptr)
+        {
+            return Marshal.PtrToStructure<T>(ptr);
+        }
+        public static void PtrToStructure<T>(System.IntPtr ptr, T structure)
+        {
+            Marshal.PtrToStructure<T>(ptr, structure);
+        }
+        public static byte ReadByte(System.IntPtr ptr)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: All Read/WRite methods should be retired.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static System.IntPtr ReadIntPtr(System.IntPtr ptr, int offset)
+        {
+            return Marshal.ReadIntPtr(ptr, offset);
+        }
+        public static System.IntPtr ReadIntPtr(object ptr, int offset)
+        {
+            return Marshal.ReadIntPtr(ptr, offset);
+        }
+        public static System.IntPtr ReallocateMemory(System.IntPtr ptr, int sizeInBytes)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: @KrzysztofCwalina @weshaggard ReallocateMemory or ReAllocateMemory?

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static int SizeOf(System.Type type)
+        {
+            return Marshal.SizeOf(type);
+        }
+        public static int SizeOf<T>()
+        {
+            return Marshal.SizeOf<T>();
+        }
+        public static System.IntPtr StringToBSTR(string s)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: Should not be here.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static System.IntPtr UnsafeAddrOfPinnedArrayElement(System.Array arr, int index)
+        {
+            return Marshal.UnsafeAddrOfPinnedArrayElement(arr, index);
+        }
+        public static System.IntPtr UnsafeAddrOfPinnedArrayElement<T>(T[] arr, int index)
+        {
+            return Marshal.UnsafeAddrOfPinnedArrayElement<T>(arr, index);
+        }
+        public static void WriteInt32(object ptr, int offset, int value)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: Should not be in PinvokeMarshal.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static System.IntPtr StringToBSTR(string s)
+        {
+            return Marshal.StringToBSTR(s);
+        }
+        public static System.IntPtr StringToAllocatedMemoryAnsi(string s)
+        {
+            return Marshal.StringToCoTaskMemAnsi(s);
+        }
+        public static System.IntPtr StringToAllocatedMemoryUni(string s)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: @weshaggard @KrzysztofCwalina There are some inconsistencies in our naming - sometimes it's Uni, sometimes it's Unicode. I think we should name it Unicode.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        public static System.IntPtr UnsafeAddrOfPinnedArrayElement<T>(T[] arr, int index)
+        {
+            return Marshal.UnsafeAddrOfPinnedArrayElement<T>(arr, index);
+        }
+        public static void WriteInt32(object ptr, int offset, int value)
+        {
+            Marshal.WriteInt32(ptr, offset, value);
+        }
+        public static void ZeroFreeBSTR(System.IntPtr s)
```
Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: BSTR should not be here.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs (hunk 19 above)
Comment (KrzysztofCwalina) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: What does "Unicode" mean in this context? Does it mean UTF16? If yes, we should call it StringToAllocatedMemoryUtf16 or something like that.

Comment (yizhang82) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: It means the same encoding as our char type. So yes, UTF16.

File: src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs (hunk 16 above)
Comment (botaberg) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: ReallocateMemory should be fine right?

Comment (weshaggard) [inline src/System.Runtime.InteropServices.PInvoke/src/System/Runtime/InteropServices/PInvokeMarshal.cs]: Yes it should be ReallocateMemory

**Comment by stephentoub** [comment]
Is this PR still valid, or was it replaced by https://github.com/dotnet-bot/corefx/commit/4af3f030a158af663dfc42533adab290b294114b ?

**Comment by weshaggard** [comment]
I think it is replaced, so closing. @botaberg if that isn't the case please submit a new PR with the differences that are needed.
